### PR TITLE
Add broadcasts to nonlinear solvers

### PR DESCRIPTION
**Pull request by devmotion**
This PR add broadcasts to non-linear solvers (see https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/pull/580#discussion_r244633802). Moreover, adds mass-matrix support to the out of place version, adds comments, and (hopefully) simplifies handling of failed steps.

### CODE REVIEW

File: src/nlsolve/functional.jl [hunk 1]
```diff
@@ -38,40 +38,48 @@ function (S::NLFunctional{false})(integrator)
...
+    end
+
+    # check early stopping criterion
     ndzprev = ndz
-    ndz = integrator.opts.internalnorm(dz)
+    ndz = integrator.opts.internalnorm(z₊ .- z)
     θ = ndz/ndzprev
-    if θ > 1 || ndz*(θ^(max_iter - iter)/(1-θ)) > κtol
-      fail_convergence = true
+    if θ ≥ 1 || ndz * θ^(max_iter - iter) > κtol * (1 - θ)
```
Comment (YingboMa) [inline src/nlsolve/functional.jl]: I am not sure if this is correct. Maybe
```julia
θ ≥ 1 || ndz * θ^( (max_iter - iter)/(1-θ) ) > κtol && break
```
?

File: src/nlsolve/functional.jl [hunk 2]
```diff
@@ -120,20 +130,21 @@ function (S::NLFunctional{true})(integrator)
     @. dz = z₊ - z
     ndzprev = ndz
     ndz = integrator.opts.internalnorm(dz)
+
+    # check early stopping criterion
     θ = ndz/ndzprev
-    if θ > 1 || ndz*(θ^(max_iter - iter)/(1-θ)) > κtol
-      fail_convergence = true
+    if θ ≥ 1 || ndz * θ^(max_iter - iter) > κtol * (1 - θ)
```
Comment (YingboMa) [inline src/nlsolve/functional.jl]: Same here.

**Comment by YingboMa** [comment]
Thanks! Could you please add some tests on solving ODEs with a mass matrix by out-of-place functional iteration?

File: src/nlsolve/functional.jl (hunk 1 above)
Comment (YingboMa) [reply src/nlsolve/functional.jl]: Sorry, never mind, I confused myself. :-)

**Comment by ChrisRackauckas** [comment]
I agree that adding those tests would be a good idea, but it can merge without it.

**Comment by ChrisRackauckas** [comment]
I opened https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/issues/586 to track the missing test, but will merge anyways.

**Pull request merged**